src/art.rs

```diff
@@ -0,0 +1,913 @@
+#[derive(Debug, Clone)]
+struct InnerNode<V> {
+    meta: Meta,               // Metadata of the node
+    node_type: NodeType,      // Type of the node
+    keys: Box<[u8]>,          // Array of keys
+    children: Box<[Node<V>]>, // Array of child nodes
+}
+
+// Leaf node of the adaptive radix trie
+#[derive(Debug, Clone)]
+struct LeafNode<V> {
+    key: Box<[u8]>, // Key associated with the leaf node
+    pub value: V,   // Value associated with the leaf node
+}
+
+// From the specification: Radix trees consist of two types of nodes:
+// Inner nodes, which map partial keys to other nodes,
+// and leaf nodes, which store the values corresponding to the keys.
+#[derive(Debug, Clone, Default)]
+enum Node<V> {
+    #[default]
+    Empty, // Empty node #[default] Empty
+    Leaf(Box<LeafNode<V>>),   // Leaf node
+    Inner(Box<InnerNode<V>>), // Inner node
+}
+
+// Adaptive radix trie
+#[derive(Debug, Clone)]
+pub struct Tree<V> {
+    root: Node<V>, // Root node of the tree
+    size: u64,     // Number of elements in the tree
+}
+
+// Default implementation for the Tree struct
+impl<V> Default for Tree<V> {
+    fn default() -> Self {
+        Tree {
+            root: Node::Empty,
+            size: 0,
+        }
+    }
+}
+
+// Implementation of methods for the Meta struct
+impl Meta {
+    // Check the length of the common prefix between the node's prefix and the given key
+    fn check_prefix(&self, key: &[u8]) -> usize {
+        let mut i = 0;
+        while i < self.prefix_len && i < key.len() && self.prefix[i] == key[i] {
+            i += 1;
+        }
+        i
+    }
+
+    // Create a new Meta instance with the given prefix, prefix length, and number of children
+    fn new(prefix: &[u8], prefix_len: usize, num_children: usize) -> Self {
+        let mut p = [0; MAX_PREFIX_LEN];
+        p[..prefix_len].copy_from_slice(&prefix[..prefix_len]);
+        Meta {
+            prefix: p,
+            prefix_len,
+            num_children,
+        }
+    }
+}
+
+// Implementation of methods for the LeafNode struct
+impl<V> LeafNode<V> {
+    // Create a new LeafNode instance with the given key and value
+    pub fn new(key: &[u8], value: V) -> Self {
+        Self {
+            key: key.into(),
+            value,
+        }
+    }
+
+    // Check if the leaf node matches the given key
+    #[inline]
+    pub fn matches(&self, key: &[u8]) -> bool {
+        if self.key.len() != key.len() {
+            return false;
+        }
+        self.key.iter().zip(key).all(|(a, b)| a == b)
+    }
+
+    // Find the length of the longest common prefix between the leaf node and another leaf node, starting from a given depth
+    #[inline]
+    pub fn longest_common_prefix(&self, other: &Self, depth: usize) -> usize {
+        let limit = min(self.key.len(), other.key.len()) - depth;
+        for idx in 0..limit {
+            if self.key[depth + idx] != other.key[depth + idx] {
+                return idx;
+            }
+        }
```

Comment:
Nitpick: This could be replaced with iterator methods:
```rust
    pub fn longest_common_prefix_2(&self, other: &Self, depth: usize) -> usize {
        self.key[depth..]
            .iter()
            .zip(&other.key[depth..])
            .position(|(a, b)| a != b)
            .unwrap_or_else(|| self.key.len().min(other.key.len()))
    }
```
I personally prefer them over loops.

Reply:
Rust codgen is also often better at optimizing iterators than for loops, so you might even get faster code.

Reply:
Thank you! Fixed this